src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,106 @@
+import axios from 'axios'
+import { log } from '../util'
+import { APPROVE_TOKEN_FOR_TRANSFER, V3_ROUTER2_ADDRESS } from '../util/callData'
+
+export const TENDERLY_BATCH_SIMULATE_API = (
+  TENDERLY_BASE_URL: string,
+  TENDERLY_USER: string,
+  TENDERLY_PROJECT: string
+) => `${TENDERLY_BASE_URL}/api/v1/account/${TENDERLY_USER}/project/${TENDERLY_PROJECT}/simulate-batch`
+
+/**
+ * Provider for dry running transactions on tenderly.
+ *
+ * @export
+ * @interface ISimulator
+ */
+export interface ISimulator {
+  /**
+   * Returns the gas fee that was paid to land the transaction in the simulation.
+   * @returns number or Error
+   */
+  simulateTransaction: (
+    chainId: number,
+    hexData: string,
+    tokenInAddress: string,
+    fromAddress: string,
+    fallback?: number
+  ) => Promise<number|Error>
+}
+export class TenderlyProvider implements ISimulator {
+  TENDERLY_BASE_URL: string
+  TENDERLY_USER: string
+  TENDERLY_PROJECT: string
+  TENDERLY_ACCESS_KEY: string
+  constructor(TENDERLY_BASE_URL: string, TENDERLY_USER: string, TENDERLY_PROJECT: string, TENDERLY_ACCESS_KEY: string) {
+    this.TENDERLY_BASE_URL = TENDERLY_BASE_URL
+    this.TENDERLY_USER = TENDERLY_USER
+    this.TENDERLY_PROJECT = TENDERLY_PROJECT
+    this.TENDERLY_ACCESS_KEY = TENDERLY_ACCESS_KEY
+  }
+
+  public async simulateTransaction(
+    chainId: number,
+    hexData: string,
+    tokenInAddress: string,
+    fromAddress: string,
+    fallback?: number,
+    stateOverrides?: Record<string, unknown>
+  ): Promise<number|Error> {
```

Comment:
This function should take a SwapRoute object, and return a new SwapRoute object (with all the gas values updated)

It also shouldn't return an Error. On error it should throw.